### PR TITLE
Tools: Remove dependencies installed by `make tools`

### PR DESCRIPTION
**Pull request by FirePing32**
Signed-off-by: Prakhar Gurunani <prakhargurunani@gmail.com>

## Description
The script removes the tools and dependencies installed by `make tools` and also removes the symlinks between them.

## Related Issue(s)

fixes: #14288

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required

### CODE REVIEW

**Comment by FirePing32** [comment]
Should this script be added somewhere in the docs?

**Comment by frouioui** [comment]
> Should this script be added somewhere in the docs?

I think it makes the most sense to me to have a new rule in the Makefile that calls this script. The same way we have `make tools`, we could have `make clean_tools`.

File: tools/remove_dependencies.sh
Comment (frouioui) [inline tools/remove_dependencies.sh:43]: It seems like the symlink on `protoc` is still present even after executing this function.

```
root@x:~/vitess# ls bin/
consul  etcd  etcdctl  protoc  toxiproxy-server  zksrv.sh

root@x:~/vitess# cd tools/

root@x:~/vitess/tools# ./remove_dependencies.sh 
## local system details...
## platform: Linux target:x86_64 OS: 
Removing protoc...
Removing zookeeper...
Removing etcd...
Removing consul...
Removing toxiproxy...

root@x:~/vitess/tools# ls ../bin/
protoc  zksrv.sh

root@x:~/vitess/tools# cd ..

root@x:~/vitess# ./bin/protoc
-bash: ./bin/protoc: No such file or directory

root@x:~/vitess# ls bin -l 
total 0
lrwxrwxrwx 1 root root 43 Oct 18 22:50 protoc -> /root/vitess/dist/vt-protoc-21.3/bin/protoc
lrwxrwxrwx 1 root root 33 Oct 18 22:52 zksrv.sh -> /root/vitess/go/vt/zkctl/zksrv.sh
```

Comment (frouioui) [reply tools/remove_dependencies.sh:43]: On another note, I think the symlink to `zksrv.sh` is fine to keep as it gets created by `build.env`

**Review by frouioui — APPROVED** [review]
Looks good to me besides that small potential issue with `protoc`'s symlink. Thank you for this contribution.

I have tested this on a remote machine with a clean clone of `vitess` running `make tools` and the new script added by this PR.

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:42]: I don't really understand this. If the passed path is a file, why are we then treating it like a directory?

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:60]: This should mean that it's not installed and we don't need to remove it. Why exit the entire uninstall?

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:71]: What's the reasoning for `unlink` in some places, `rm` in others, and `rm -f` in yet others?

**Review by mattlord — COMMENTED** [review]
I think that this is a good idea. I'm curious where you get the 9GiB assessment from? 

The files are here: https://github.com/vitessio/vitess-resources/releases/

And at least for me: 
```
❯ du -csh dist
 74M	dist
 74M	total
```

Do you really see the disk usage go down by ~ 9GB after running this script? If so, then I agree that it makes it all the more useful 🙂

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:27]: There's no need to call `name` and `uname -m` in so many places. We can set globals here:
PLATFORM=$(uname -s)
ARCH=$(uname -m)

File: tools/remove_dependencies.sh
Comment (FirePing32) [reply tools/remove_dependencies.sh:3]: That 9 Gib was a mistake actually. It also included the Go cache.

File: tools/remove_dependencies.sh
Comment (FirePing32) [reply tools/remove_dependencies.sh:42]: Yeah, that was a mistake, it should be `-f "$dist/bin/protoc"` instead.

File: tools/remove_dependencies.sh
Comment (FirePing32) [reply tools/remove_dependencies.sh:71]: unlink was just removing the symlinks and not the binaries in `$VTROOT/bin`. Had to manually remove them using `rm`. And to clean up the dist dir for each tool, used `rm -rf`.

File: tools/remove_dependencies.sh
Comment (FirePing32) [reply tools/remove_dependencies.sh:32]: I think that line is no longer needed as from now onwards this file will be invocated by `Makefile`

File: build.env
Comment (mattlord) [inline build.env:31]: Why allow custom values only for ZK? I think in this case we want them to be fixed as we pull from our repo.

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:39]: This method is fine, but you can also do this more efficiently in bash this way:
```
if [[ ${dir##*/} == "vt-protoc-$PROTOC_VER" ]]; then
```

This also removes an external dependency (awk isn't quite the same between the BSD and GNU toolchains e.g.).

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:48]: Same note as above.

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:61]: Nit, but ellipses is 3 dots (same below).

**Review by mattlord — COMMENTED** [review]
Nice work on this, @FirePing32 ! I only had a few minor comments. Let me know what you think and I can come back and re-review quickly at any time.

File: bootstrap.sh
Comment (mattlord) [inline bootstrap.sh:237]: To be overly safe, we should quote all of the VER parameters.

File: build.env
Comment (FirePing32) [reply build.env:31]: `ZK_VER` had custom values in `bootstrap.sh`. So I did the same here.

File: tools/remove_dependencies.sh
Comment (FirePing32) [reply tools/remove_dependencies.sh:29]: Yeah, sounds good.

File: tools/remove_dependencies.sh
Comment (FirePing32) [reply tools/remove_dependencies.sh:39]: [basename](https://linux.die.net/man/1/basename) is also an option.

Comment (FirePing32) [reply tools/remove_dependencies.sh]: @mattlord Should I implement `basename`? It's been a long time since I opened this PR.

Comment (mattlord) [reply tools/remove_dependencies.sh]: Any reason not to use the bash feature shown? That will be the most efficient thing and eliminates an unnecessary external dependency for the script. It also looks like we'll need to merge in origin/main to address some test failures as well.

Comment (FirePing32) [reply tools/remove_dependencies.sh]: No specific reason as such. Basename is also widely supported. We can go with your bash approach.

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:33]: Nit, but these are local vars so would typically be lowercase.

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:39]: I saw your previous commit that changed these but now don't see them here for some reason.

File: tools/remove_dependencies.sh
Comment (mattlord) [inline tools/remove_dependencies.sh:122]: We missed one spot here. Otherwise LGTM! Once we fix this I'll approve and then merge assuming all tests pass.

**Comment by mattlord** [comment]
@FirePing32 Sorry, one last thing. We should address the shellcheck warnings — these should have been surfaced by the pre-commit hook `.git/hooks/pre-commit` which then ends up running `misc/git/hooks/shellcheck`. They are all minor variable quoting issues:
```
❯ shellcheck tools/remove_dependencies.sh

In tools/remove_dependencies.sh line 40:
        rm -rf $dist
               ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
        rm -rf "$dist"


In tools/remove_dependencies.sh line 49:
        rm -rf $dist
               ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
        rm -rf "$dist"


In tools/remove_dependencies.sh line 60:
        Darwin) local platform=darwin; local ext=zip;;
                                             ^-^ SC2034 (warning): ext appears unused. Verify use (or export if used externally).


In tools/remove_dependencies.sh line 81:
        rm -rf $dist
               ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
        rm -rf "$dist"


In tools/remove_dependencies.sh line 94:
        rm -rf $dist
               ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
        rm -rf "$dist"


In tools/remove_dependencies.sh line 122:
        rm -rf $dist
               ^---^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
        rm -rf "$dist"


In tools/remove_dependencies.sh line 139:
    uninstall_etcd $ETCD_VER "$VTROOT/dist/etcd"
                   ^-------^ SC2086 (info): Double quote to prevent globbing and word splitting.

Did you mean:
    uninstall_etcd "$ETCD_VER" "$VTROOT/dist/etcd"

For more information:
  https://www.shellcheck.net/wiki/SC2034 -- ext appears unused. Verify use (o...
  https://www.shellcheck.net/wiki/SC2086 -- Double quote to prevent globbing ...
```

If we don't do this then it will show up for other people when modifying related files.

**Comment by mattlord** [comment]
Just one left: 
```
❯ shellcheck tools/remove_dependencies.sh

In tools/remove_dependencies.sh line 60:
        Darwin) local platform=darwin; local ext=zip;;
                                             ^-^ SC2034 (warning): ext appears unused. Verify use (or export if used externally).

For more information:
  https://www.shellcheck.net/wiki/SC2034 -- ext appears unused. Verify use (o...
```

Did you intend to use that `$ext` variable or should we remove it?

**Comment by FirePing32** [comment]
I guess we don't need that now.

**Comment by mattlord** [comment]
As part of a minimal manual test, I did this:
```
make tools

make clean_tools
```

`make clean_tools` fails:
```
❯ make clean_tools
make: source: No such file or directory
make: *** [clean_tools] Error 1
```

We cannot use `source` as the `Makefile` is not a shell script:
```
diff --git a/Makefile b/Makefile
index b3198536f7..43b6906087 100644
--- a/Makefile
+++ b/Makefile
@@ -393,7 +393,7 @@ tools:
        ./bootstrap.sh

 clean_tools:
diff --git a/Makefile b/Makefile
index b3198536f7..22193f8b4b 100644
--- a/Makefile
+++ b/Makefile
@@ -393,7 +393,7 @@ tools:
        ./bootstrap.sh

 clean_tools:
-       source build.env
+       bash ./build.env
        ./tools/remove_dependencies.sh

 minimaltools:
```

But then it still fails:
```
❯ make clean_tools
## local system details...
## platform: Darwin target:arm64 OS: darwin23
./tools/remove_dependencies.sh: line 131: PROTOC_VER: unbound variable
make: *** [clean_tools] Error 1
```

And even when manually sourcing it:
```
❯ make clean_tools
## local system details...
## platform: Darwin target:arm64 OS: darwin23
Removing protoc...
Removing zookeeper...
Removing etcd...
rm: /Users/matt/git/vitess/bin/etcd: No such file or directory
make: *** [clean_tools] Error 1
```

Can you please run some basic manual tests to ensure that this is working as intended?

**Comment by mattlord** [comment]
For the main issue, you could do this or use `${VTROOT}/build.env` if you prefer:
```
diff --git a/Makefile b/Makefile
index b3198536f7..625a28baac 100644
--- a/Makefile
+++ b/Makefile
@@ -393,7 +393,6 @@ tools:
        ./bootstrap.sh

 clean_tools:
-       source build.env
        ./tools/remove_dependencies.sh

 minimaltools:
diff --git a/tools/remove_dependencies.sh b/tools/remove_dependencies.sh
index 32a6f56496..5609a5e230 100755
--- a/tools/remove_dependencies.sh
+++ b/tools/remove_dependencies.sh
@@ -16,6 +16,8 @@

 # Remove tools and dependencies installed by "make tools"

+source "$(dirname "${BASH_SOURCE[0]:-$0}")/../build.env"
+
 set -euo pipefail

 function fail() {
```

**Review by mattlord — APPROVED** [review]
Thank you for the contribution, @FirePing32 ! I tested it and it all seems to be working as expected. ❤️

**Comment by mattlord** [comment]
I'm going to override and merge because this new code is not part of any tests -- and there are two tests which are out to pasture with the Expected ... state. The only thing you can do in those cases is push another empty commit but that is not needed in this case.

**Pull request merged**
